[ERROR] [TEARDOWN]: Connection already closed - mysql_conn.store_result() - MySQLdb._exceptions.OperationalError: (2006, '')

### What action do you want to perform
I would like to test with pytest-mysql a function that connects to the mysqldb (patched/mocked) and closes the connection in the function.

### What are the results
Testing with pytest-mysql I get this exception that turns out into 2 unexpected-behaviour:
```
request = <SubRequest 'mysqlnoproc_client' for <Function test_userExists_dummy_with_old_connect>>

    @pytest.fixture
    def mysql_fixture(request: FixtureRequest) -> MySQLdb.Connection:
        """
        Client fixture for MySQL server.
    
        #. Get config.
        #. Try to import MySQLdb package.
        #. Connect to mysql server.
        #. Create database.
        #. Use proper database.
        #. Drop database after tests.
    
        :param request: fixture request object
    
        :returns: connection to database
        """
        config = get_config(request)
        process: Union[
            NoopMySQLExecutor, MySQLExecutor
        ] = request.getfixturevalue(process_fixture_name)
        if not process.running():
            process.start()
    
        mysql_user = process.user
        mysql_passwd = passwd or config["passwd"]
        mysql_db = dbname or config["dbname"]
    
        connection_kwargs = {
            "host": process.host,
            "user": mysql_user,
            "passwd": mysql_passwd,
        }
        if process.unixsocket:
            connection_kwargs["unix_socket"] = process.unixsocket
        else:
            connection_kwargs["port"] = process.port
    
        query_str = (
            f"CREATE DATABASE {mysql_db} "
            f"DEFAULT CHARACTER SET {charset} "
            f"DEFAULT COLLATE {collation}"
        )
        try:
            mysql_conn: MySQLdb.Connection = _connect(
                connection_kwargs, query_str
            )
        except OperationalError:
            # Fallback to mysql connection with root user
            connection_kwargs["user"] = "root"
            mysql_conn: MySQLdb.Connection = _connect(
                connection_kwargs, query_str
            )
        mysql_conn.query(f"USE {mysql_db}")
        yield mysql_conn
    
        # clean up after test that forgot to fetch selected data
>       mysql_conn.store_result()
E       MySQLdb._exceptions.OperationalError: (2006, '')

../../../venv/lib/python3.8/site-packages/pytest_mysql/factories/client.py:134: OperationalError
```
So --> mysql_conn.store_result() is not closed in a try/except block and so --> function test gets all right but test fails with the exception raised.
Also this is not good with a connection mysql_noproc because database doesn't get DROPPED and we need to go into the console and drop it manually


### What are the expected results
 - [x] Try/except block surrounding mysql_conn.store_result()
 - [x] Database dropped anyway
 - [x] If connection is closed at mysql_conn.store_result() --> do nothing

If this makes sense to some people, I'll try propose a PR with these changes

--- a/src/pytest_mysql/factories/client.py
+++ b/src/pytest_mysql/factories/client.py
@@ -58,7 +58,9 @@ def mysql(
     :rtype: func
     """
 
-    def _connect(connect_kwargs: dict, query_str: str, mysql_db: str) -> MySQLdb.Connection:
+    def _connect(
+            connect_kwargs: dict, query_str: str, mysql_db: str
+    ) -> MySQLdb.Connection:
         """Apply given query to a  given MySQLdb connection."""
         mysql_conn: MySQLdb.Connection = MySQLdb.connect(**connect_kwargs)
         try:
@@ -132,7 +134,9 @@ def mysql(
 
         # clean up after test that forgot to fetch selected data
         if not mysql_conn.open:
-            mysql_conn: MySQLdb.Connection = MySQLdb.connect(**connection_kwargs)
+            mysql_conn: MySQLdb.Connection = MySQLdb.connect(
+                **connection_kwargs
+            )
         try:
             mysql_conn.store_result()
         except Exception as e:
@@ -140,4 +144,5 @@ def mysql(
         query_drop_database = "DROP DATABASE IF EXISTS %s" % mysql_db
         mysql_conn.query(query_drop_database)
         mysql_conn.close()
+
     return mysql_fixture

--- a/src/pytest_mysql/factories/client.py
+++ b/src/pytest_mysql/factories/client.py
@@ -131,8 +131,27 @@ def mysql(
         yield mysql_conn
 
         # clean up after test that forgot to fetch selected data
-        mysql_conn.store_result()
-        mysql_conn.query("DROP DATABASE IF EXISTS %s" % mysql_db)
-        mysql_conn.close()
+        try:
+            if mysql_conn.open:
+                mysql_conn.store_result()
+        except Exception as e:
+            print(str(e))
+        try:
+            if not mysql_conn.open:
+                try:
+                    mysql_conn: MySQLdb.Connection = _connect(
+                        connection_kwargs, query_str
+                    )
+                except OperationalError:
+                    # Fallback to mysql connection with root user
+                    connection_kwargs["user"] = "root"
+                    mysql_conn: MySQLdb.Connection = _connect(
+                        connection_kwargs, query_str
+                    )
+                mysql_conn.query(f"USE {mysql_db}")
+            mysql_conn.query("DROP DATABASE IF EXISTS %s" % mysql_db)
+            mysql_conn.close()
+        except Exception as e:
+            print(str(e))
 
     return mysql_fixture

--- a/src/pytest_mysql/factories/client.py
+++ b/src/pytest_mysql/factories/client.py
@@ -137,19 +137,21 @@ def mysql(
         except Exception as e:
             print(str(e))
         try:
+            query_drop_database = "DROP DATABASE IF EXISTS %s" % mysql_db
             if not mysql_conn.open:
                 try:
                     mysql_conn: MySQLdb.Connection = _connect(
-                        connection_kwargs, "DROP DATABASE IF EXISTS %s" % mysql_db
+                        connection_kwargs,
+                        query_drop_database
                     )
                 except OperationalError:
                     # Fallback to mysql connection with root user
                     connection_kwargs["user"] = "root"
                     mysql_conn: MySQLdb.Connection = _connect(
-                        connection_kwargs, "DROP DATABASE IF EXISTS %s" % mysql_db
-                    )
+                        connection_kwargs,
+                        query_drop_database)
             else:
-                mysql_conn.query("DROP DATABASE IF EXISTS %s" % mysql_db)
+                mysql_conn.query(query_drop_database)
             mysql_conn.close()
         except Exception as e:
             print(str(e))

--- a/src/pytest_mysql/factories/client.py
+++ b/src/pytest_mysql/factories/client.py
@@ -58,7 +58,7 @@ def mysql(
     :rtype: func
     """
 
-    def _connect(connect_kwargs: dict, query_str: str) -> MySQLdb.Connection:
+    def _connect(connect_kwargs: dict, query_str: str, mysql_db: str) -> MySQLdb.Connection:
         """Apply given query to a  given MySQLdb connection."""
         mysql_conn: MySQLdb.Connection = MySQLdb.connect(**connect_kwargs)
         try:
@@ -66,7 +66,7 @@ def mysql(
         except ProgrammingError as e:
             if "database exists" in str(e):
                 raise DatabaseExists(
-                    f"Database {query_str} already exists. There's some test "
+                    f"Database {mysql_db} already exists. There's some test "
                     f"configuration error. Either you start your own server "
                     f"with the database name used in tests, or you use two "
                     f"fixtures with the same database name on the same "
@@ -119,41 +119,25 @@ def mysql(
         )
         try:
             mysql_conn: MySQLdb.Connection = _connect(
-                connection_kwargs, query_str
+                connection_kwargs, query_str, mysql_db
             )
         except OperationalError:
             # Fallback to mysql connection with root user
             connection_kwargs["user"] = "root"
             mysql_conn: MySQLdb.Connection = _connect(
-                connection_kwargs, query_str
+                connection_kwargs, query_str, mysql_db
             )
         mysql_conn.query(f"USE {mysql_db}")
         yield mysql_conn
 
         # clean up after test that forgot to fetch selected data
+        if not mysql_conn.open:
+            mysql_conn: MySQLdb.Connection = MySQLdb.connect(**connection_kwargs)
         try:
-            if mysql_conn.open:
-                mysql_conn.store_result()
+            mysql_conn.store_result()
         except Exception as e:
             print(str(e))
-        try:
-            query_drop_database = "DROP DATABASE IF EXISTS %s" % mysql_db
-            if not mysql_conn.open:
-                try:
-                    mysql_conn: MySQLdb.Connection = _connect(
-                        connection_kwargs,
-                        query_drop_database
-                    )
-                except OperationalError:
-                    # Fallback to mysql connection with root user
-                    connection_kwargs["user"] = "root"
-                    mysql_conn: MySQLdb.Connection = _connect(
-                        connection_kwargs,
-                        query_drop_database)
-            else:
-                mysql_conn.query(query_drop_database)
-            mysql_conn.close()
-        except Exception as e:
-            print(str(e))
-
+        query_drop_database = "DROP DATABASE IF EXISTS %s" % mysql_db
+        mysql_conn.query(query_drop_database)
+        mysql_conn.close()
     return mysql_fixture

--- a/src/pytest_mysql/factories/client.py
+++ b/src/pytest_mysql/factories/client.py
@@ -140,16 +140,16 @@ def mysql(
             if not mysql_conn.open:
                 try:
                     mysql_conn: MySQLdb.Connection = _connect(
-                        connection_kwargs, ""
+                        connection_kwargs, "DROP DATABASE IF EXISTS %s" % mysql_db
                     )
                 except OperationalError:
                     # Fallback to mysql connection with root user
                     connection_kwargs["user"] = "root"
                     mysql_conn: MySQLdb.Connection = _connect(
-                        connection_kwargs, ""
+                        connection_kwargs, "DROP DATABASE IF EXISTS %s" % mysql_db
                     )
-                mysql_conn.query(f"USE {mysql_db}")
-            mysql_conn.query("DROP DATABASE IF EXISTS %s" % mysql_db)
+            else:
+                mysql_conn.query("DROP DATABASE IF EXISTS %s" % mysql_db)
             mysql_conn.close()
         except Exception as e:
             print(str(e))

--- a/src/pytest_mysql/factories/client.py
+++ b/src/pytest_mysql/factories/client.py
@@ -66,7 +66,7 @@ def mysql(
         except ProgrammingError as e:
             if "database exists" in str(e):
                 raise DatabaseExists(
-                    f"Database {mysql_db} already exists. There's some test "
+                    f"Database {query_str} already exists. There's some test "
                     f"configuration error. Either you start your own server "
                     f"with the database name used in tests, or you use two "
                     f"fixtures with the same database name on the same "
@@ -140,13 +140,13 @@ def mysql(
             if not mysql_conn.open:
                 try:
                     mysql_conn: MySQLdb.Connection = _connect(
-                        connection_kwargs, query_str
+                        connection_kwargs, ""
                     )
                 except OperationalError:
                     # Fallback to mysql connection with root user
                     connection_kwargs["user"] = "root"
                     mysql_conn: MySQLdb.Connection = _connect(
-                        connection_kwargs, query_str
+                        connection_kwargs, ""
                     )
                 mysql_conn.query(f"USE {mysql_db}")
             mysql_conn.query("DROP DATABASE IF EXISTS %s" % mysql_db)

--- a/src/pytest_mysql/factories/client.py
+++ b/src/pytest_mysql/factories/client.py
@@ -59,7 +59,7 @@ def mysql(
     """
 
     def _connect(
-            connect_kwargs: dict, query_str: str, mysql_db: str
+        connect_kwargs: dict, query_str: str, mysql_db: str
     ) -> MySQLdb.Connection:
         """Apply given query to a  given MySQLdb connection."""
         mysql_conn: MySQLdb.Connection = MySQLdb.connect(**connect_kwargs)

--- a/tests/test_mysqlnoproc.py
+++ b/tests/test_mysqlnoproc.py
@@ -11,3 +11,17 @@ def test_mysql_noproc(mysqlnoproc_client):
     cursor.execute(QUERY)
     mysqlnoproc_client.commit()
     cursor.close()
+
+
+def test_mysql_noproc_closing_connection_not_throwing_exception(
+    mysqlnoproc_client,
+):
+    """
+    Check if closing the connection doesn't throw an exception
+    when cleaning the fixture.
+    """
+    cursor = mysqlnoproc_client.cursor()
+    cursor.execute(QUERY)
+    mysqlnoproc_client.commit()
+    cursor.close()
+    mysqlnoproc_client.close()

